### PR TITLE
adding undo button

### PR DESCRIPTION
**Pull request by govind007**
Screenshot 2023-09-10 at 4 56 23 PM

### CODE REVIEW

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 1]
```diff
@@ -20,6 +21,30 @@ export default function WithoutOptions({
...
 }: Props) {
+  const [userResponse, setUserResponse] = useState({
+    isConfirmed: false,
+    value: null as any,
+  });
+  const [timer, setTimer] = useState(0);
+  const timerRef = useRef(null as any);
+
+  useEffect(() => {
+    if (timer > 0) {
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: This check is redundant

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 2]
```diff
@@ -20,6 +21,30 @@ export default function WithoutOptions({
...
+      if (timer === 5) {
+        clearTimeout(timerRef.current);
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: Instead of doing this here execute it in useEffect return statement so that time is cleared when the component unmounts

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 3]
```diff
@@ -28,6 +53,24 @@ export default function WithoutOptions({
...
+    setUserResponse({
+      isConfirmed: true,
+      value: choice,
+    });
+    setTimer(timer + 1);
+  }
+
+  function undoSubmit() {
+    clearTimeout(timerRef.current);
+    timerRef.current = null;
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: No need to do this

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 4]
```diff
@@ -48,15 +91,28 @@ export default function WithoutOptions({
         sx={(theme) => getQuestionTextStyles(theme, !!options[0].text && options[0].isCorrect)}>
         <SanitizedHtml>{options[0].text}</SanitizedHtml>
       </Box>
-      {!isAttempted && (
+      {userResponse.isConfirmed ? (
         <Group>
-          <Button color="red" onClick={() => handleSelectedChoice([])}>
-            Incorrect
+          <Button variant="outline" onClick={() => handleSelectedChoice(userResponse.value)}>
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: No need to click this button. Automatically do this when timer is up. Also, remove this button.

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 5]
```diff
@@ -28,6 +53,24 @@ export default function WithoutOptions({
...
+      isConfirmed: true,
+      value: choice,
+    });
+    setTimer(timer + 1);
+  }
+
+  function undoSubmit() {
+    clearTimeout(timerRef.current);
+    timerRef.current = null;
+    setTimer(0);
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: No need to reset the timer. As it will run only once after mount. On remount it will reset automatically. Also, I was thinking if we should do countdown timer in state. Start from 5 and keep decrementing till 0 and when 0 execute submit.

File: src/components/QuestionPlay/index.tsx [hunk 6]
```diff
@@ -116,9 +116,11 @@ export default function QuestionPlay({
           />
         )}
         {isAttempted && !isGameCompleted && (
-          <Button mt="xl" variant="default" onClick={continueGame}>
-            Continue
-          </Button>
+          <>
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/index.tsx]: Why this change? Please revert.

File: src/components/QuestionPlay/styles.module.css [hunk 7]
```diff
@@ -1,4 +1,33 @@
...
+}
+
+@keyframes countdown {
+  from {
+    stroke-dashoffset: 0px;
+  }
+  to {
+    stroke-dashoffset: 113px;
+  }
+}
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/styles.module.css]: Move this to a separate file undoTimer.module.css

File: src/components/QuestionPlay/undoTimer.tsx [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/undoTimer.tsx]: Capital case component file name

Comment (rajatkantinandi) [reply src/components/QuestionPlay/undoTimer.tsx]: The component name should be UndoButtonIcon?

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 9]
```diff
@@ -20,6 +21,30 @@ export default function WithoutOptions({
...
+  const [timer, setTimer] = useState(0);
+  const timerRef = useRef(null as any);
+
+  useEffect(() => {
+    if (timer > 0) {
+      if (timer === 5) {
+        clearTimeout(timerRef.current);
+        handleSelectedChoice(userResponse.value);
+        setUserResponse({
+          isConfirmed: false,
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: This is odd. When timer is up isConfirmed is set to false?

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 10]
```diff
@@ -20,6 +21,30 @@ export default function WithoutOptions({
   setIsAnswerRevealed,
   setSelectedChoices,
 }: Props) {
+  const [userResponse, setUserResponse] = useState({
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: isConfirmed is confusing. Instead, we should just use isUndoTimerRunning boolean and store userResponse in a ref so that it can be submitted later.

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 1 above)
Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx]: We should check isUndoTimerRunning

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 11]
```diff
@@ -28,6 +53,24 @@ export default function WithoutOptions({
     }, 100);
   }
 
+  function handleSubmitResponse(choice: number[]) {
+    setUserResponse({
+      isConfirmed: true,
+      value: choice,
+    });
+    setTimer(timer + 1);
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: We will decrement after 1s not now. Just set isUndoTimerRunning = true and store the ref of userResponse.

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 12]
```diff
@@ -48,15 +91,28 @@ export default function WithoutOptions({
         sx={(theme) => getQuestionTextStyles(theme, !!options[0].text && options[0].isCorrect)}>
         <SanitizedHtml>{options[0].text}</SanitizedHtml>
       </Box>
-      {!isAttempted && (
+      {userResponse.isConfirmed ? (
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: use isUndoTimerRunning

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 13]
```diff
@@ -28,6 +53,24 @@ export default function WithoutOptions({
...
+      value: choice,
+    });
+    setTimer(timer + 1);
+  }
+
+  function undoSubmit() {
+    clearTimeout(timerRef.current);
+    timerRef.current = null;
+    setTimer(0);
+    setUserResponse({
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: set isUndoTimerRunning to false here.

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 9 above)
Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx]: Set isUndoTimerRunning to false here.

**Review by rajatkantinandi — CHANGES_REQUESTED** [review]
This isConfirmed logic is confusing and Confirm button is not required. Added comments to cleanup the code.

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 2 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: we have to stop timer when either user confirmed the answer or 5 secs are over. nothing to with component unmount

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 1 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: done

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 4 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: if there are many subjective questions (or if the user is doing a quick testing for his own quiz), it would be frustrating for user to wait for 5 sec after each response

Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx]: They don't need to wait for 5s. We automatically save it after 5s. Even if they click another question. I think better to have the timer outside of the component in a helper function.

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 2 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: we are unmounting the component when user click on continue button

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 4 above)
Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx]: I have a better idea. Show the continue button instead of this button. Clicking continue will save the response and stop timer, right?

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 11 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: I'm showing current time in undo button, that is why storing in state
Screenshot 2023-10-20 at 12 23 43 PM

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 2 above)
Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx]: Yeah, clear it when 5s are over as well as if it unmounts when user clicks continue while the timer is running.

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 11 above)
Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx]: OK

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 5 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: I'm showing countdown timer in `Undo` button

File: src/components/QuestionPlay/index.tsx (hunk 6 above)
Comment (govind007) [reply src/components/QuestionPlay/index.tsx]: reverted

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 4 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: Screenshot 2023-10-20 at 1 17 30 PM
Screenshot 2023-10-20 at 1 17 37 PM

This is the current UI, user can't click another question before submitting response

Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: showing 'Continue' button instead of confirm Submit

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 2 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: fixed

File: src/components/QuestionPlay/UndoButtonIcon.tsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import styles from './undoButtonIcon.module.css';
+
+export default function UndoTimer({ children }) {
+  return (
+    <>
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/UndoButtonIcon.tsx]: Why don't render the button here itself?

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 15]
```diff
@@ -18,14 +20,50 @@ export default function WithoutOptions({
...
-      document.getElementById('btnSubmitResponse')?.click();
-    }, 100);
+  useEffect(() => {
+    if (isUndoTimerRunning) {
+      timerRef.current = setTimeout(() => {
+        if (time === 4) {
+          handleContinueClick(userResponse.current);
+        }
+
+        setTime(time + 1);
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: We don't need to do this when time === 4? Maybe put this in else?

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 16]
```diff
@@ -18,14 +20,50 @@ export default function WithoutOptions({
...
+  }, [isUndoTimerRunning, time]);
+
+  function handleContinueClick(choice: number[]) {
+    submitResponse(choice);
+    continueGame();
+  }
+
+  function handleSubmitResponse(choice: number[]) {
+    if (isUndoClicked) {
+      submitResponse(choice);
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx:56]: Submit response when undo button clicked? Feels odd.

Comment (rajatkantinandi) [reply src/components/QuestionPlay/WithoutOptions.tsx:56]: Is this `wasUndoClickedOnce`?

File: src/components/QuestionPlay/WithoutOptions.tsx [hunk 17]
```diff
@@ -18,14 +20,50 @@ export default function WithoutOptions({
...
+  const userResponse = useRef(null as any);
 
-    setTimeout(() => {
-      document.getElementById('btnSubmitResponse')?.click();
-    }, 100);
+  useEffect(() => {
+    if (isUndoTimerRunning) {
+      timerRef.current = setTimeout(() => {
+        if (time === 4) {
+          handleContinueClick(userResponse.current);
```
Comment (rajatkantinandi) [inline src/components/QuestionPlay/WithoutOptions.tsx]: Why do we click continue automatically? We should just hide the undo button here and stop timer. Let user click continue as they might have not finished seeing the result.

File: src/styles/theme.ts [hunk 18]
```diff
@@ -58,7 +58,7 @@ const theme: MantineThemeOverride = {
         icon: params.color
           ? {}
           : {
-              color: 'var(--qm-primary)',
+              color: 'var(--qm-primary) !important',
```
Comment (rajatkantinandi) [inline src/styles/theme.ts]: Why this change?

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 16 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx:56]: yes.. changed the name

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 17 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: fixed

File: src/components/QuestionPlay/WithoutOptions.tsx (hunk 15 above)
Comment (govind007) [reply src/components/QuestionPlay/WithoutOptions.tsx]: fixed

**Review by rajatkantinandi — APPROVED** [review]
